src/magic_folder/test/test_upload.py

```diff
@@ -0,0 +1,164 @@
+        )
+
+    def get_local_snapshot(self, path, _unused):
+        for (name, snapshot) in self.local_processed:
+            if name is path:
+                return snapshot
+        return None
+
+    def remove_localsnapshot(self, path):
+        for p in self.local_processed:
+            (name, snapshot) = p
+            if name is path:
+                self.local_processed.remove(p)
+
+    def store_remote_snapshot(self, path, cap):
+        self.remote_processed.append((path, cap))
+
+    def get_remote_snapshot_cap(self, path):
+        for (p, remote_snapshot) in self.remote_processed:
+            if p is path:
+                return remote_snapshot.content_cap
+
+class UploaderServiceTests(SyncTestCase):
+    """
+    Tests for ``UploaderService``.
+    """
+    def setUp(self):
+        super(UploaderServiceTests, self).setUp()
+        self.magic_path = FilePath(self.mktemp())
+        self.magic_path.makedirs()
+        self.author = create_local_author("alice")
+        self.stash_dir = FilePath(mktemp())
+        self.stash_dir.makedirs()
+
+    def setup_example(self):
+        """
+        per-example state that is invoked by hypothesis.
+        """
+        self.root = create_fake_tahoe_root()
+        self.http_client = create_tahoe_treq_client(self.root)
+        self.tahoe_client = create_tahoe_client(
+            DecodedURL.from_text(u"http://example.com"),
+            self.http_client,
+        )
+        self.snapshot_store = MemorySnapshotStore()
+        self.uploader_service = UploaderService(
+            snapshot_store=self.snapshot_store,
+            local_author = self.author,
+            tahoe_client=self.tahoe_client,
+            clock=reactor,
+            polling_interval=1,
+        )
+
+    @given(path_segments(), binary())
+    def test_commit_a_file(self, name, content):
+        """
+        Add a file into localsnapshot store, start the service which
+        should result in a remotesnapshot corresponding to the
+        localsnapshot.
+        """
+
+        # create a local snapshot
+        data = io.BytesIO(content)
+
+        d = create_snapshot(
+            name=name,
+            author=self.author,
+            data_producer=data,
+            snapshot_stash_dir=self.stash_dir,
+            parents=[],
+        )
+
+        # push LocalSnapshot object into the SnapshotStore.
+        d.addCallback(self.snapshot_store.store_local_snapshot)
+
+        # start Uploader Service
+        self.uploader_service.startService()
+
+        # this should be picked up by the Uploader Service and should
+        # result in a snapshot cap.
+        d.addCallback(lambda _unused:
+                      self.snapshot_store.get_remote_snapshot_cap(name))
+
+        # test whether we got a capability
+        self.assertThat(
+            d,
+            succeeded(
+                AfterPreprocessing(is_uri, Always()),
+            ),
+        )
+
+        self.uploader_service.stopService()
```

Comment:
If the assertion fails this code never gets run.  In this case, it's nice if the stopService cleanups get a chance to happen anyway so a nice pattern is:

```
self.uploader_service.startService()
self.addCleanup(self.uploader_service.stopService)

...
```

"Cleanups" added this way always run even if the test fails.